### PR TITLE
Emergency Network restart

### PR DESCRIPTION
**Pull request by Leo-Besancon**
Needs testing

* [x] document all added functions
* [x] try in sandbox /simulation/labnet
* [x] unit tests on the added/changed features
  * [x] make tests compile
  * [x] make tests pass 
* [ ] add logs allowing easy debugging in case the changes caused problems
* [ ] if the API has changed, update the API specification

### CODE REVIEW

File: massa-pos-exports/src/pos_final_state.rs
Comment (Ben-PH) [inline massa-pos-exports/src/pos_final_state.rs:111]: there's a few clones in here. can it be done without by any chance?

if partial move of self is a problem, one possibility is:

```rust
let Self { 
    /// fields go here
} = self;
```
that way, you don't have to use `self.$FIELD` anymore, and helps get around partial moves

File: massa-pos-exports/src/pos_final_state.rs
Comment (Ben-PH) [inline massa-pos-exports/src/pos_final_state.rs:120]:
```rust
let iter = deferred_credits.credits.clone()
    .iter()
    .filter(|(slot, _)| slot > from_slot) // even better: if this is known to be sorted, do a binary search
    .take_while(|(slot, _)| slot <= end_slot);
for (slot, map) in iter {
    for ... in map.into_iter() { 
        ....
    }
}
```

Then look into using flatten so that you only have one `for`, which goes straight into the the iterator over all (addr, amount)s

File: massa-pos-exports/src/pos_final_state.rs
Comment (Ben-PH) [inline massa-pos-exports/src/pos_final_state.rs:619]: Why not make the feilds public, and remove the setters?

File: massa-pos-worker/src/worker.rs
Comment (Ben-PH) [inline massa-pos-worker/src/worker.rs:123]:
```rust
let Command::DrawInput{
    // fields
} = self.input_mpsc.recv() else {break;};
```

File: massa-bootstrap/src/client.rs
Comment (Ben-PH) [inline massa-bootstrap/src/client.rs:469]: move this one to the top (it's the shorter return scope)


add context to each of these scopes so that the next reader can easily see there's a series of early-return conditions

**Review by Ben-PH — COMMENTED** [review]
Nicely done! Only one or two comments that are meaningful, the rest are style/nit/personal preference.

A couple of things that would be repeated comments:

- struct fields doc-commented to just be the name of the field being commented. In these cases, write a comment that provides context if it is deserved, or just replace the comment with an `#[allow(...)]` for that field/struct
- there's the trio of values that appear tegother regularly: rolls, seed, last-start-period. See if they deserve to make up a dedicated struct.
- have a look at using the `pinned_cf` methods of rocks-db, and use at your discretion.

File: massa-pos-exports/src/pos_final_state.rs
Comment (Leo-Besancon) [reply massa-pos-exports/src/pos_final_state.rs]: Done in f67dd591f21f38de7095af9fccd1cb66c4c2fe1d (+ combined the fields in InitialState struct)

File: massa-final-state/src/final_state.rs
Comment (Leo-Besancon) [reply massa-final-state/src/final_state.rs:68]: Done in f67dd591f21f38de7095af9fccd1cb66c4c2fe1d

File: massa-executed-ops/src/executed_ops.rs
Comment (Leo-Besancon) [reply massa-executed-ops/src/executed_ops.rs:70]: Done in f67dd591f21f38de7095af9fccd1cb66c4c2fe1d

File: massa-bootstrap/src/messages.rs
Comment (Leo-Besancon) [reply massa-bootstrap/src/messages.rs]: Done in f67dd591f21f38de7095af9fccd1cb66c4c2fe1d

File: massa-execution-worker/src/worker.rs
Comment (Leo-Besancon) [reply massa-execution-worker/src/worker.rs:64]: I think we only need it here

File: massa-models/src/block_header.rs
Comment (Leo-Besancon) [reply massa-models/src/block_header.rs]: Done in f67dd591f21f38de7095af9fccd1cb66c4c2fe1d

File: massa-bootstrap/src/messages.rs
Comment (Ben-PH) [reply massa-bootstrap/src/messages.rs]: nice

File: massa-executed-ops/src/executed_ops.rs
Comment (Ben-PH) [reply massa-executed-ops/src/executed_ops.rs:70]: nice

File: massa-final-state/src/final_state.rs
Comment (Leo-Besancon) [reply massa-final-state/src/final_state.rs:120]: Done in eec62b33db173755bf5c609be58f15c018ee1728, let me know what you think about it :)

File: massa-final-state/src/final_state.rs
Comment (Leo-Besancon) [reply massa-final-state/src/final_state.rs]: Removed that method in 0bdbc013ea7d909577d261e5573acf3325d78691, it is handled differently, directly in the finalize() method

File: massa-final-state/src/final_state.rs
Comment (Leo-Besancon) [reply massa-final-state/src/final_state.rs]: The finalize() method is called in the execution thread with a write lock, so it is not necessary: nothing can change the final_state apart from the execution module at that point.

File: massa-pos-exports/src/pos_final_state.rs
Comment (Leo-Besancon) [reply massa-pos-exports/src/pos_final_state.rs]: Used `BTreeMap`'s `range` method to iterate in da77c706d49425ce526501a0a2facdcc923bf4df.
However, I still need the slot variable in the inner loop, so I couldn't make the flatten idea work. Maybe you have an idea?

Comment (Ben-PH) [reply massa-pos-exports/src/pos_final_state.rs]: BtreeMap iteration would probably be less performant than a sort => binary-search => take-until: collecting a btreemap is O(n log_n), same as the sort, but iterating through a btreemap is not nearly as clean as iterating through a vector (tree-traversal and cache misses), but this is only relevant if this is a performance hot-spot.

The issue of readability is what prompted this comment initially: The filter and take_while methods will handle the need for the `continue` and `break` checks, and the thinking was the `flatten` to remove the nested `for`.

using the filter/bin-search + take_while should be easy enough, and make the code easier to parse for the next person to come along. The `flatten` benefit to readability is debatable, though.

File: massa-pos-exports/src/pos_final_state.rs
Comment (Ben-PH) [reply massa-pos-exports/src/pos_final_state.rs:113]: Righto.

File: massa-models/src/block.rs
Comment (Ben-PH) [reply massa-models/src/block.rs]: I was predicting an `allow` would be chosen, but I like this even better. Nice.

File: massa-final-state/src/final_state.rs
Comment (Ben-PH) [reply massa-final-state/src/final_state.rs:120]: Pretty happy with it.

There are a lot of clones of the config. Consider Rc-ing it, or using references instead of moves if this function could be a performance burden. If it's rarely called, leave as is

Comment (Leo-Besancon) [reply massa-final-state/src/final_state.rs:120]: It should never be called hopefully haha. Called once, on 1 node, for each emergency restart we have :)

File: massa-final-state/src/final_state.rs
Comment (Ben-PH) [reply massa-final-state/src/final_state.rs]: Nice. Added some comments to the changes, but they can be addressed seperately

File: massa-final-state/src/final_state.rs
Comment (Ben-PH) [reply massa-final-state/src/final_state.rs:120]: ...where's the Oprah meme "you get a clone! you get a clone! everyone gets a clone!" when you need it!

File: massa-ledger-worker/src/ledger_db.rs
Comment (Ben-PH) [reply massa-ledger-worker/src/ledger_db.rs:445]: ...given this is a rarely used, non-repeating function, and on having another look, using `get_pinned...` slice-returning methods could be expensive, not worth the effort I think.

File: massa-ledger-worker/src/ledger_db.rs
Comment (Ben-PH) [inline massa-ledger-worker/src/ledger_db.rs]: If it's close enough to trivial for the db_iterator to get slices instead of vecs (is there an `iterator_pinned_cf`?), then that might be nice here, but don't spend too much time on it.

Comment (Ben-PH) [reply massa-ledger-worker/src/ledger_db.rs]: My bad. Didn't realize github interface took me to an older commit.

File: massa-final-state/src/final_state.rs
Comment (Ben-PH) [inline massa-final-state/src/final_state.rs:200]: this is duplicated in the other if.  latest_cycle_info can be defined just the once, I think

File: massa-final-state/src/final_state.rs
Comment (Ben-PH) [inline massa-final-state/src/final_state.rs:192]: move the two bodies of this if-statement into two helper functions, the names of which help self-document


(e.g. `interpolate_single_cycle` and `interpolate_multiple_cycles`, or something along those lines)

File: massa-final-state/src/final_state.rs
Comment (Ben-PH) [inline massa-final-state/src/final_state.rs:247]: the `last_slot` creation inside the for loop can be done in the iterator creation itself. This suggested change (not to be commited directly) forms a template such that the body of this loop is strictly about behavior, without any value declaration distractions. This should allow better readability for future maintainers>

```suggestion
            // We now need to visit all the slots in the already completed cycles... 
            let completed_slots_iter = (current_slot_cycle + 1..end_slot_cycle).map(|cycle| Slot::new_last_of_cycle(cycle, ...).map_err(|err| {
                    FinalStateError::InvalidSlot(format!(
                        "Cannot create slot for interpolating downtime: {}",
                        err
                    ))
                });
            // ...and at each slot, we need to by $404-reason-not-found
            for last_lot in slot_iter {
            
```

File: massa-final-state/src/final_state.rs
Comment (Ben-PH) [inline massa-final-state/src/final_state.rs:262]: consider using helper functions and namings that help in self-documenting what is happening here.

File: massa-ledger-worker/src/ledger_db.rs
Comment (Ben-PH) [inline massa-ledger-worker/src/ledger_db.rs:176]: Consider using included/excluded without the saturating sub (similar idea to using `0..thread_count` instead of `0..=(thread_count - 1)`

File: massa-ledger-worker/src/ledger_db.rs
Comment (Ben-PH) [inline massa-ledger-worker/src/ledger_db.rs:505]: propogate error instead of panic?

File: massa-ledger-worker/src/ledger_db.rs
Comment (Ben-PH) [inline massa-ledger-worker/src/ledger_db.rs:509]: same db call made twice?

File: massa-models/src/config/constants.rs
Comment (Ben-PH) [inline massa-models/src/config/constants.rs:125]: I'm not sure if this change was a decision as part of this feature, or to aid in testing. Prompt to ensure this PR lands with correct periods/cycle count.

File: massa-node/src/main.rs
Comment (Ben-PH) [inline massa-node/src/main.rs:172]: Might be good to rename so that it more clearly conveys that something is being created (in this case, using last_start_period as a seed-value in a way), as opposed to some sort of `from_$TYPE` conversion method.

```suggestion
            FinalState::derive_from_period(
```

File: massa-pos-exports/src/pos_final_state.rs
Comment (Ben-PH) [inline massa-pos-exports/src/pos_final_state.rs]: Why are these tests removed?

**Review by Ben-PH — COMMENTED** [review]
Nice work.

Potential sticking points are the removal of tests, but aside from that, it's mostly just prompts and trivial (I hope) readability changes.

Please go through and check for unwraps and expects, and propogate errors instead of panicing as appropriate.

File: massa-models/src/config/constants.rs
Comment (Leo-Besancon) [reply massa-models/src/config/constants.rs:125]: Yes it's for testing (avoid waiting hours in sandox / labnet just to get a CycleUnavailable :D), I'll add a comment

Comment (Ben-PH) [reply massa-models/src/config/constants.rs]: keep it unresolved for now, just so it doesn't slip through. This comment can serve as a checklist-item

File: massa-final-state/src/final_state.rs
Comment (Leo-Besancon) [reply massa-final-state/src/final_state.rs:247]: I'm not sure about this one, I'll try to refactor to make it clearer (maybe with helper functions and comments), but IMO it's clearer to thing about: 
> "I need to create the cycle N, and for that I need info on the last_slot of cycle N"

than to think about

> "I'll extend the cycle_info to cover everything until last_slot, that conveniently is the last slot of a cycle I'm trying to build"

File: massa-final-state/src/final_state.rs
Comment (Leo-Besancon) [reply massa-final-state/src/final_state.rs:262]: Done in refactor : 1405887a97bd4fbf5444e4b09f5cf2da48919a27

File: massa-final-state/src/final_state.rs
Comment (Leo-Besancon) [reply massa-final-state/src/final_state.rs:180]: Done in refactor 1405887a97bd4fbf5444e4b09f5cf2da48919a27

File: massa-final-state/src/final_state.rs
Comment (Leo-Besancon) [reply massa-final-state/src/final_state.rs]: Now that it is used in two different calls, it seems a bit weird to pass around this mut (as I already have a &mut self), see refactor 1405887a97bd4fbf5444e4b09f5cf2da48919a27

File: massa-final-state/src/final_state.rs
Comment (Leo-Besancon) [reply massa-final-state/src/final_state.rs]: Added a comment in a2f16fbfeebe3ff858d0e5cc4d7764e3e2b5b4ea

File: massa-pos-exports/src/pos_final_state.rs
Comment (Leo-Besancon) [reply massa-pos-exports/src/pos_final_state.rs]: The test no longer be needed when I edit the way deferred credits are handled, but it was a premature removal, sorry! See https://github.com/massalabs/massa/pull/3740

File: massa-node/src/main.rs
Comment (Leo-Besancon) [reply massa-node/src/main.rs:172]: I don't think the last_start_period variable is that important here. Yes the value is useful to interpolate the downtime, but the primary goal is to recover all the values from the snapshot!

Maybe new_derived_from_snapshot()?

File: massa-ledger-worker/src/ledger_db.rs
Comment (Leo-Besancon) [reply massa-ledger-worker/src/ledger_db.rs:505]: I think it's appropriate here because that's totally unrecoverable (the ledger was corrupted), and is done the same way for all other rocks_db calls.

If you think strongly about this maybe we should open an issue on a refactor around this? (E.g. create a new error type to be propagated for all db call fails?)

File: massa-ledger-worker/src/ledger_db.rs
Comment (Leo-Besancon) [reply massa-ledger-worker/src/ledger_db.rs:509]: Good catch :)
Fixed in 5596b96fea113585dfab2b4ee7dc5165ac044db4

File: massa-ledger-worker/src/ledger_db.rs
Comment (Leo-Besancon) [reply massa-ledger-worker/src/ledger_db.rs]: Makes sense, thanks! Done in 850eac71d0b6228262e323cfcf595109ec266ca8

File: massa-models/src/config/constants.rs
Comment (Leo-Besancon) [reply massa-models/src/config/constants.rs]: Resolved now

File: massa-final-state/src/final_state.rs
Comment (Ben-PH) [reply massa-final-state/src/final_state.rs]: Hmmm. This feels like a dilemma. On the one hand, it is a repeated code, but on the other hand, it does seem to make sense to be in the helpers.

Might make sens to make a helper method to make the latest cycle info, but ill leave that to your discression.

File: massa-final-state/src/final_state.rs
Comment (Ben-PH) [reply massa-final-state/src/final_state.rs]: Ideally, you could do something like this

`(....).try_map($MAKE_SLOT)?.for_each($DO_POS_STATE_THINGS)`

...bat that's non-trivial without doing the whole construction first.

If this is a common loop, and pos-state activities are expensive, it might be worth considering building all the slots first anyway, to make sure it's a valid exercise. The idea being is that you don't want to get to the very end, having done all that, and fail.

...but it's a discression thing. This isn't worth spending too much time or energy on.

File: massa-ledger-worker/src/ledger_db.rs
Comment (Ben-PH) [reply massa-ledger-worker/src/ledger_db.rs:502]: I'm happy with that.

File: massa-ledger-worker/src/ledger_db.rs
Comment (Ben-PH) [reply massa-ledger-worker/src/ledger_db.rs]: nice.

File: massa-node/src/main.rs
Comment (Ben-PH) [reply massa-node/src/main.rs:172]: That makes sense, go for it. I'll leave it to you to resolve this thread to mark that change as done.

File: massa-ledger-worker/src/ledger_db.rs
Comment (Ben-PH) [reply massa-ledger-worker/src/ledger_db.rs]: nice. 

Come to think of it, try`0` instead of `u8::MIN`?

Once you've given it thought, go with whichever you think is better, then go ahead and resolve-mark this thread

File: massa-pos-exports/src/pos_final_state.rs
Comment (Ben-PH) [reply massa-pos-exports/src/pos_final_state.rs]: Cheers. We can resolve this thread once that PR is brought in.

Comment (Leo-Besancon) [reply massa-pos-exports/src/pos_final_state.rs]: Updated test in d7aa30b57ed4c97223a55b55a728087c559f46db

File: massa-final-state/src/final_state.rs
Comment (Leo-Besancon) [reply massa-final-state/src/final_state.rs]: As it's ran only once, I'll let it as is right now. However, I keep the note for the follow-up next month ;)

File: massa-models/src/test_exports/tools.rs
Comment (Ben-PH) [inline massa-models/src/test_exports/tools.rs:18]: less than 0?

File: massa-network-exports/src/settings.rs
Comment (Ben-PH) [inline massa-network-exports/src/settings.rs:96]: Provide context (or replace comment with a clippy-allow)

File: massa-node/src/main.rs
Comment (Ben-PH) [inline massa-node/src/main.rs:183]:
```suggestion
    // Create final state, either from a snapshot, or from scratch
```

File: massa-node/src/main.rs
Comment (Ben-PH) [inline massa-node/src/main.rs:201]: consider putting the `Arc::new(parking_lot::RwLock::new(` around the match-statement

```rust
let final_state = Arc::new(parking_lot::RwLock::new(match ... {
    Some(last....) => FinalState::new_....(
    ....
     ) ,
    None => FinalState::new(
    ....
    )
}));
```

File: massa-pos-exports/src/deferred_credits.rs
Comment (Ben-PH) [inline massa-pos-exports/src/deferred_credits.rs:31]: Why is it optional?

File: massa-pos-exports/src/deferred_credits.rs
Comment (Ben-PH) [inline massa-pos-exports/src/deferred_credits.rs:62]: "get the hash resulting from"... add context

File: massa-pos-exports/src/pos_final_state.rs
Comment (Ben-PH) [inline massa-pos-exports/src/pos_final_state.rs:322]: if this can be pub(crate), make it so.

**Review by Ben-PH — COMMENTED** [review]
Congratulations. This is looking gooood. For the purpose of testing Testnet 21 before deployment, feel free to consider this PR approved by me.

In places I've said "consider". Even if I don't explicitly say so, feel free to resolve these comments yourself, with or without changes applied.

Apart from some nits, I couldn't find anything. I don't think I'd be able to catch anything that Damir would miss when it comes to the design details and the correctness of implementation.

There are a lot of places where doc-comments don't provide any more context than what is provided by the function/field in its names and/or types. In these cases, it's my personal opinion that such comments add redundant noise. Please consider providing useful context for the next person to come along, or use the appropriate `#[allow(clippy::(...))]`

File: massa-final-state/src/error.rs
Comment (Leo-Besancon) [reply massa-final-state/src/error.rs:19]: This one is useful for displaydoc: https://docs.rs/displaydoc/latest/displaydoc/

File: massa-models/src/test_exports/tools.rs
Comment (Leo-Besancon) [reply massa-models/src/test_exports/tools.rs:18]: This was from the time I wanted to use last_start_period here, but I'm not, so I'll reuse the `==` :)

File: massa-pos-exports/src/pos_final_state.rs
Comment (Leo-Besancon) [reply massa-pos-exports/src/pos_final_state.rs:322]: I have to use it when interpolating downtime in final_state, so I sadly can't

File: massa-final-state/src/final_state.rs
Comment (Leo-Besancon) [reply massa-final-state/src/final_state.rs:240]: I think I'll clean the downtime interpolation in the follow-up, I have borrow-checker issues doing it naively here.

**Review by Ben-PH — APPROVED** [review]

**Pull request merged**
